File: src/AutoContext.Worker.Web/src/hosting/worker-protocol-channel.ts
import type { Duplex } from 'node:stream';

/**
 * Wire-protocol adapter for the worker pipe: 4-byte little-endian
 * payload length followed by that many UTF-8 JSON bytes. Wraps a
 * Node {@link Duplex} stream (typically a `net.Socket`) and exposes
 * message-oriented read/write operations on top of it.
 *
 * This is the contract counterpart of the C# `WorkerProtocolChannel` in
 * `AutoContext.Worker.Shared`; the two implementations are
 * bit-for-bit symmetric and must be changed together.
 *
 * The wrapped stream's lifetime is owned by the caller — this class
 * neither ends nor destroys it.
 */
export class WorkerProtocolChannel {
    /**
     * Maximum payload size accepted by {@link WorkerProtocolChannel.read}.
     * Caps allocation when a corrupted or hostile header arrives;
     * tasks exchanged on this pipe are small JSON envelopes well
     * below this limit.
     */
    static readonly MAX_MESSAGE_BYTES = 64 * 1024 * 1024;

    private readonly stream: Duplex;

    constructor(stream: Duplex) {
        this.stream = stream;
    }

    /**
     * Reads one length-prefixed message from the wrapped stream.
     * Returns `null` when the connection is closed before a full
     * header is received. Throws on header overflow or when
     * cancellation is signaled mid-read.
     */
    async read(signal?: AbortSignal): Promise<Buffer | null> {
        const header = await this.readExact(4, signal);

        if (header === null) {
            return null;
        }

        const length = header.readInt32LE(0);

        if (length < 0) {
            throw new Error(`Pipe message length ${length} is negative; header is corrupt.`);
        }

        if (length === 0) {
            return Buffer.alloc(0);
        }

        if (length > WorkerProtocolChannel.MAX_MESSAGE_BYTES) {
            throw new Error(
                `Pipe message length ${length} exceeds the maximum of ${WorkerProtocolChannel.MAX_MESSAGE_BYTES} bytes.`,
            );
        }

        return this.readExact(length, signal);
    }

    /**
     * Writes {@link payload} to the wrapped stream as one
     * length-prefixed message.
     */
    async write(payload: Buffer, signal?: AbortSignal): Promise<void> {
        throwIfAborted(signal);

        const message = Buffer.allocUnsafe(4 + payload.length);
        message.writeInt32LE(payload.length, 0);
        payload.copy(message, 4);

        const stream = this.stream;
        await new Promise<void>((resolve, reject) => {
            let abortListener: (() => void) | undefined;

            const cleanup = (): void => {
                if (abortListener !== undefined) {
                    signal?.removeEventListener('abort', abortListener);
                }
            };

            if (signal !== undefined) {
                abortListener = (): void => {
                    cleanup();
                    reject(signalAbortError(signal));
                };
                signal.addEventListener('abort', abortListener, { once: true });
            }

            stream.write(message, (err) => {
                cleanup();
                if (err !== null && err !== undefined) {
                    reject(err);
                    return;
                }
                resolve();
            });
        });
    }

    private readExact(byteCount: number, signal: AbortSignal | undefined): Promise<Buffer | null> {
        const stream = this.stream;
        return new Promise((resolve, reject) => {
            if (signal?.aborted === true) {
                reject(signalAbortError(signal));
                return;
            }

            if (byteCount === 0) {
                resolve(Buffer.alloc(0));
                return;
            }

            let settled = false;

            const settle = (action: () => void): void => {
                if (settled) {
                    return;
                }
                settled = true;
                stream.removeListener('readable', onReadable);
                stream.removeListener('end', onEnd);
                stream.removeListener('error', onError);
                if (signal !== undefined) {
                    signal.removeEventListener('abort', onAbort);
                }
                action();
            };

            const tryRead = (): void => {
                if (settled) {
                    return;
                }
                // `read(n)` returns exactly `n` bytes when available, or
                // null otherwise — avoiding the unshift/re-entry pitfalls
                // of the flowing 'data' mode. At EOF with fewer than `n`
                // bytes buffered, Node returns the partial remainder; we
                // treat that as a clean "stream ended" outcome.
                const chunk = stream.read(byteCount) as Buffer | null;
                if (chunk === null) {
                    return;
                }
                if (chunk.length < byteCount) {
                    settle(() => resolve(null));
                    return;
                }
                settle(() => resolve(chunk));
            };

            const onReadable = (): void => {
                tryRead();
            };

            const onEnd = (): void => {
                settle(() => resolve(null));
            };

            const onError = (err: Error): void => {
                settle(() => reject(err));
            };

            const onAbort = (): void => {
                settle(() => reject(signalAbortError(signal!)));
            };

            stream.on('readable', onReadable);
            stream.on('end', onEnd);
            stream.on('error', onError);
            if (signal !== undefined) {
                signal.addEventListener('abort', onAbort, { once: true });
            }

            // Data may already be buffered (common for back-to-back reads
            // on the same stream); try once before relying on events.
            tryRead();
        });
    }
}

function throwIfAborted(signal: AbortSignal | undefined): void {
    if (signal?.aborted === true) {
        throw signalAbortError(signal);
    }
}

function signalAbortError(signal: AbortSignal): Error {
    const reason: unknown = signal.reason;
    if (reason instanceof Error) {
        return reason;
    }
    const err = new Error('The operation was aborted.');
    err.name = 'AbortError';
    return err;
}

File: src/AutoContext.Framework/Workers/WorkerProtocolChannel.cs
namespace AutoContext.Framework.Workers;

using System.Buffers.Binary;

/// <summary>
/// Wire-protocol adapter for the worker pipe: 4-byte little-endian
/// payload length followed by that many UTF-8 JSON bytes. Wraps an
/// underlying <see cref="Stream"/> (typically a named-pipe stream) and
/// exposes message-oriented read/write operations on top of it.
/// </summary>
/// <remarks>
/// This is the contract counterpart of <c>WorkerProtocolChannel</c> in
/// <c>AutoContext.Worker.Web</c>; the two implementations are
/// bit-for-bit symmetric and must be changed together.
///
/// The wrapped stream's lifetime is owned by the caller — this type
/// neither closes nor disposes it.
/// </remarks>
public sealed class WorkerProtocolChannel
{
    /// <summary>
    /// Maximum payload size accepted by <see cref="ReadAsync"/>.
    /// Caps allocation when a corrupted or hostile header arrives;
    /// tasks exchanged on this pipe are small JSON envelopes well
    /// below this limit.
    /// </summary>
    public const int MaxMessageBytes = 64 * 1024 * 1024; // 64 MiB

    private readonly Stream _stream;

    /// <summary>
    /// Wraps <paramref name="stream"/> with the pipe wire protocol.
    /// The stream is not owned by this instance.
    /// </summary>
    public WorkerProtocolChannel(Stream stream)
    {
        ArgumentNullException.ThrowIfNull(stream);

        _stream = stream;
    }

    /// <summary>
    /// Reads one length-prefixed message from the wrapped stream.
    /// Returns <see langword="null"/> when the connection is closed
    /// before a full header is received.
    /// </summary>
    public async Task<byte[]?> ReadAsync(CancellationToken ct)
    {
        var header = new byte[4];
        var headerRead = await ReadExactAsync(header, ct).ConfigureAwait(false);

        if (!headerRead)
        {
            return null;
        }

        var length = BinaryPrimitives.ReadInt32LittleEndian(header);

        if (length < 0)
        {
            throw new InvalidDataException(
                $"Pipe message length {length} is negative; header is corrupt.");
        }

        if (length == 0)
        {
            return [];
        }

        if (length > MaxMessageBytes)
        {
            throw new InvalidDataException(
                $"Pipe message length {length} exceeds the maximum of {MaxMessageBytes} bytes.");
        }

        var payload = new byte[length];
        var payloadRead = await ReadExactAsync(payload, ct).ConfigureAwait(false);

        return payloadRead ? payload : null;
    }

    /// <summary>
    /// Writes <paramref name="payload"/> to the wrapped stream as one
    /// length-prefixed message.
    /// </summary>
    public async Task WriteAsync(byte[] payload, CancellationToken ct)
    {
        ArgumentNullException.ThrowIfNull(payload);

        var message = new byte[4 + payload.Length];
        BinaryPrimitives.WriteInt32LittleEndian(message, payload.Length);
        payload.CopyTo(message.AsSpan(4));

        await _stream.WriteAsync(message, ct).ConfigureAwait(false);
        await _stream.FlushAsync(ct).ConfigureAwait(false);
    }

    private async Task<bool> ReadExactAsync(byte[] buffer, CancellationToken ct)
    {
        var offset = 0;

        while (offset < buffer.Length)
        {
            var read = await _stream.ReadAsync(
                buffer.AsMemory(offset, buffer.Length - offset), ct).ConfigureAwait(false);

            if (read == 0)
            {
                return false;
            }

            offset += read;
        }

        return true;
    }
}
